src/Network/ABCI/Types/App.hs

```diff
@@ -1,21 +1,89 @@
+import qualified Network.ABCI.Types.DecodeError             as DecodeError
 import           Network.ABCI.Types.Messages.Request  (Request)
+import qualified Network.ABCI.Types.Messages.Request  as Request
 import           Network.ABCI.Types.Messages.Response (Response)
+import qualified Network.ABCI.Types.Messages.Response as Response
 import           Network.ABCI.Types.Messages.Types    (MessageType)
+import qualified Proto.Types                          as PT
+import qualified Proto.Types_Fields                   as PT
 
 -- | Application type that represents a well typed application, i.e. a
 -- function from a typed `Request` to a typed `Response`.
 newtype App m = App
-  { runApp :: forall (t :: MessageType). Request t -> m (Response t) }
+  { unApp :: forall (t :: MessageType). Request t -> m (Response t) }
 
 -- | Transform an application from running in a custom monad to running in `IO`.
-transformApp
-  :: (forall a. m a -> n a)
-  -> App m
-  -> App n
-transformApp nat (App f) = App $ \req ->
-  nat $ f req
+transformApp :: (forall a. m a -> g a) -> App m -> App g
+transformApp nat (App f) = App $ nat . f
+
+-- | compiles `App` down to `AppBS`
+runApp :: Applicative m => App m -> AppBS m
+runApp (App app) bs =
+  bs
+    & (decodeLengthPrefix >=> decodeRequests)
+    & either (pure . onError) (traverse onResponce)
+    & fmap (encodeLengthPrefix . encodeResponces)
+  where
+    onError :: DecodeError -> [PT.Response]
+    onError err = [Response.toProto $ Response.ResponseException $ Response.Exception $ cs $ DecodeError.print err]
+    -- for some reason with this type we get • Couldn't match type ‘m1’ with ‘m’
+    -- onResponce :: PT.Request'Value -> m PT.Response
+    onResponce = Request.withProto $ fmap Response.toProto . app
+
+-- | ByteString which contains multiple length prefixed messages
+type LengthPrefixedByteString = BS.ByteString
+
+-- | compiled Application which operates on varlength-prefixed ByteString
+type AppBS m = LengthPrefixedByteString -> m LengthPrefixedByteString
+
+
+-- | Encodes responces to bytestrings
+encodeResponces :: [PT.Response] -> [BS.ByteString]
+encodeResponces = map PL.encodeMessage
+
+-- | Decodes bytestrings into requests
+decodeRequests :: [BS.ByteString] -> Either DecodeError [PT.Request'Value]
+decodeRequests = traverse $ \packet -> case PL.decodeMessage packet of
+  Left parseError -> Left $ DecodeError.CanNotDecodeRequest packet parseError
+  Right (request :: PT.Request) -> case request ^. PT.maybe'value of
+    Nothing -> Left $ DecodeError.NoValueInRequest packet (request ^. PL.unknownFields)
+    Just value -> Right $ value
+
+-- | Encodes ByteStrings into varlength-prefixed ByteString
+encodeLengthPrefix :: [BS.ByteString] -> LengthPrefixedByteString
+encodeLengthPrefix = foldMap $ \bytes ->
+  let headerN = signedInt64ToWord . fromIntegral . BS.length $ bytes
+      header = runBuilder $ putVarInt headerN
+  in header `BS.append` bytes
+{-# INLINEABLE encodeLengthPrefix #-}
```

Comment:
Do we still need this INLINEABLE ? if yes shuold we add it to other functions too?

Reply:
removed